### DISCUSSION TITLE
Logging is sometimes duplicated without color

### DISCUSSION THREAD

**Discussion by thomasaarholt** [borntyping/python-colorlog#118] [Support]
Asking here primarily to see if you know an obvious reason why this is happening:

I'm sometimes getting a double logging output in my log:
```python
1-19 14:26:36 DEBUG    First dates from partial predictions: {'transfer_rates': datetime.date(2020, 10, 5), 'orders_day': datetime.date(2020, 10, 5), 'units_order': datetime.date(2020, 10, 5)}
11-19 14:26:36 DEBUG    First date of joint prediction: 2020-10-05
11-19 14:26:36 DEBUG    Days predicted in partial predictions fron joint first date: {'transfer_rates': 42, 'orders_day': 42, 'units_order': 42}
11-19 14:26:36 DEBUG    Last date of joint prediction: 2020-11-15
11-19 14:26:38 DEBUG    Using expected new [[ 850. 1200. 1200. 1200. 3000. 3000. 3000.]]
11-19 14:26:38 DEBUG    Using expected winback [[2400. 2600. 2600. 2500. 3000. 3000. 3200.]]
11-19 14:26:38 DEBUG    Added noise to expected new and winback
11-19 14:26:38 DEBUG    Normalize ranges and draws on prediction for transfer_rates
DEBUG:main:Normalize ranges and draws on prediction for transfer_rates
11-19 14:26:38 DEBUG    Normalize ranges and draws on prediction for orders_day
DEBUG:main:Normalize ranges and draws on prediction for orders_day
11-19 14:26:38 DEBUG    Normalize ranges and draws on prediction for units_order
DEBUG:main:Normalize ranges and draws on prediction for units_order
11-19 14:26:38 INFO     Add adjustments
INFO:main:Add adjustments
```
![image](https://user-images.githubusercontent.com/2721423/142640393-b972c4e3-d1f9-4ef6-89ca-d10b0f296da6.png)

The white lines containing :main: are the lines I do not expect. Once these lines start appearing, the double logs appear with every single logging statement.

We import a logger in our scripts that is defined as:

```python
def make_logger(console=True):
    logger = logging.getLogger(name)
    logger.setLevel(level)

    if console:
        color_fmt = "%(log_color)s" if use_color and console else ""
        # some additional unrelated code here, removed for clarity
   
        # set logger stream variable so color only in interactive
        # terminals AND notebooks, preventing ANSI escape signatures
        # in grayscale log files
        stream = None if is_in_ipython_session() else sys.stderr

        formatter = colorlog.ColoredFormatter(
            fmt=color_fmt + log_format_for_console,
            datefmt=time_format_for_console,
            log_colors={
                "DEBUG": "blue",
                "INFO": "green",
                "WARNING": "yellow",
                "ERROR": "red",
                "CRITICAL": "bold_red,bg_white",
            },
            stream=stream,
        )
        if log_in_gmt:
            formatter.converter = time.gmtime

        handler = colorlog.StreamHandler()
        handler.setFormatter(formatter)

        logger.addHandler(handler)
```

I've use this as `logger = make_logger()`, and then e.g. `logger.debug(...)`. I've noticed that I can trigger the double line behaviour if I accidentally call `logging.debug` instead of `logger.debug`, but that is not the case in the above.

Do you have any idea what's going on?

**Comment by borntyping** [comment]
It's extremly likely that somewhere in the code another handler is being created. Maybe `make_logger` is being called twice?

**Comment by thomasaarholt** [comment]
You might be right, although I'm struggling to figure out where. I'll try to raise an error if a handler already exists, and see what the traceback says!

**Comment by thomasaarholt** [comment]
This might be related. I notice that if I erroneously call `logging.debug` instead of `logger.debug`, all future `logger.debug` calls are printed twice, second time without color. 

```python
import colorlog
import logging

handler = colorlog.StreamHandler()
handler.setFormatter(colorlog.ColoredFormatter(
	'%(log_color)s%(levelname)s:%(name)s:%(message)s'))

logger = colorlog.getLogger('example')
logger.addHandler(handler)

logger.warning('Normal logging')
logging.warning("Logging using root") # using logging instead of logger
logger.warning('This message is printed twice') # logger now outputs message twice
```
![image](https://user-images.githubusercontent.com/2721423/142848798-6e2ca17f-6c35-4f3c-8b38-ae02e7de28f0.png)

This also applies to the normal logger:

```python
import logging

handler = logging.StreamHandler()
handler.setFormatter(logging.Formatter(
	'%(levelname)s:%(name)s:%(message)s'))

logger = logging.getLogger('example')
logger.addHandler(handler)

logger.warning('Normal logging')
# WARNING:example:Normal logging

logging.warning("Logging using root") # using logging instead of logger
# WARNING:root:Logging using root

logger.warning('This message is printed twice') # logger now outputs message twice
# WARNING:example:This message is printed twice
# WARNING:example:This message is printed twice
```

Now, I don't call logging.debug by accident (I mean, I *did* and that is how I discovered this, but that is no longer the case), but the above *does* look identical to my behaviour, except that in my case there is never output a _root_ logging message.

**Comment by thomasaarholt** [comment]
After calling the above, there is still only one handler in `logger.handlers` :/

**Comment by borntyping** [comment]
`logging.debug` sets up a handler if there are no existing handler on the `root` logger (i.e. top level logger returned by `logging.getLogger()` with no arguments). Colorlog implements something very similar in it's own `colorlog.debug` type functions.

https://github.com/python/cpython/blob/3.10/Lib/logging/__init__.py#L2137-L2139

In your example you'll probably find that `logging.getLogger().handlers != logging.getLogger('example').handlers`.

**Comment by marade** [comment]
I've encountered this bug when using the polars-bio module, which has its own logging:

https://biodatageeks.org/polars-bio/api/#polars_bio.set_loglevel

And then when I subsequently configure colorlog logging I get the duplicates, e.g.

INFO:root:Getting data.
[2026-02-10 18:12:48.211 INFO] Getting data.

My question is how do I get polars-bio to not log (or only do WARN logging) while still being able to do INFO logging with colorlog? Minimal example below. Try moving the pb.set_loglevel call before and after the colorlog configuration.

```
#!/usr/bin/env python3

import os, sys, logging, colorlog
import polars as pl
import polars_bio as pb

pb.set_loglevel("warning")

log_colors = {
    'DEBUG': 'light_white',
    'INFO': 'cyan',
    'WARNING': 'yellow',
    'ERROR': 'red',
    'CRITICAL': 'bold_red',
}

logger = colorlog.getLogger()
logger.setLevel(logging.INFO)
formatter = colorlog.ColoredFormatter('%(log_color)s[%(asctime)s.%(msecs)03d %(levelname)s] ' +\
    '%(message)s', log_colors=log_colors, datefmt='%Y-%m-%d %H:%M:%S')
stdout_handler = logging.StreamHandler(sys.stdout)
stdout_handler.setLevel(logging.INFO)
stdout_handler.setFormatter(formatter)
logger.addHandler(stdout_handler)

def main():
    df1 = pl.DataFrame([
        ['chr1', 1, 5],
        ['chr1', 3, 8],
        ['chr1', 8, 10],
        ['chr1', 12, 14]],
        schema=['chrom', 'start', 'end'],
        orient="row"
    )
    df1.config_meta.set(coordinate_system_zero_based = False) # 1-based coordinates

    df2 = pl.DataFrame([
        ['chr1', 4, 8],
        ['chr1', 10, 11]],
        schema=['chrom', 'start', 'end' ],
        orient="row"
    )
    df2.config_meta.set(coordinate_system_zero_based = False) # 1-based coordinates

    countsdf = pb.count_overlaps(df1, df2, output_type="polars.DataFrame")
    print(countsdf)
    logger.info(os.path.basename(__file__) + " done.")
    return 0

if __name__ == "__main__":
   sys.exit(main())
```

**Comment by borntyping** [comment]
It looks like `pb.set_loglevel("warning")` calls `logging.basicConfig` ([`polars_bio/logging.py`](https://github.com/biodatageeks/polars-bio/blob/55efb680135f3af4de0f5be5ff1f8ed03e620be9/polars_bio/logging.py)), which causes the duplicated output.

You can probably set the log level for it directly instead of calling `pb.set_loglevel`:

```
logging.getLogger("polars_bio").setLevel(logging.WARNING)
```

**Comment by marade** [comment]
Unfortunately no matter where I put logging.getLogger("polars_bio").setLevel(logging.WARNING) in the minimal example, I still get the duplicated output. I'll go ahead and cross-post this on the polars-bio repo Issues.
